### DISCUSSION TITLE
.filter() questions...    code brakes page when do onKeyUp

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#38618] [Help] [answered]
Say...
https://github.com/Elindo586/nextprojectb/blob/main/pages/inventory/testinventory.js

^^ On this page here I have pagination working OK...   but I want to add filtering and I would appreciate any help...

I'll write the code and I'll ask questions below it.. 

```
import React from "react";
import { useEffect, useState, useMemo } from "react";
import Head from "next/head";
import Link from "next/link";
import Pagination from "../components/pagination";

import db from "../../utils/parts-charlynn.json";

export async function getStaticProps() {
  
  return {
    props: { db },
  };
}
// console.log({ db });
const TestInventory = ({ db }) => {
  let PageSize = 10;

  const [currentPage, setCurrentPage] = useState(1);
  const [searchTerm, setSearchTerm] = useState("");

  const currentTableData = useMemo(() => {
    const firstPageIndex = (currentPage - 1) * PageSize;
    const lastPageIndex = firstPageIndex + PageSize;
    return db.slice(firstPageIndex, lastPageIndex);
  }, [currentPage]);
```
then...

```
return (
    <div>
      <div className="col-md-12">
        <span className=" invy-text">
          <p>
            stuff here
          </p>
        </span>
      </div>
      <div className="col-md-12">
        <input
          type="text"
          id="myInput"
          // onKeyUp="myFunction()"
          onKeyUp={(event) => {
            setSearchTerm(event.target.value);
          }}
          placeholder="Search for Part Number.."
          title="Type in a name"
        />
        <br />
        <br />
        <table className="my-talbe-sort table table-hover table-striped table-bordered">
          <thead>
            <tr>
              <th>Reference</th>
              <th>Description</th>
              <th>Replacement</th>
            </tr>
          </thead>
          <tbody>
            {currentTableData
              .filter((val) => {
                if (searchTerm == "") {
                  return val;
                } else if (
                  val.Reference.toLowerCase().includes(searchTerm.toLowerCase())
                ) {
                }
              })
              .map((item) => {
                return (
                  <tr>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
              })}
          </tbody>
        </table>
        <Pagination
          className="pagination-bar"
          currentPage={currentPage}
          totalCount={db.length}
          pageSize={PageSize}
          onPageChange={(page) => setCurrentPage(page)}
        />
      </div>
    </div>
  );
};
```

Well... the code breaks as soon as I do type try to type anything in the input box...

I want to .filter currentTableData and return with filtered data from all 3 columns, so far I am only doing 1 column.....  I guess I can add another || item.Description || item.Replacements to the same .filter line of code....  ???

Also...  `totalCount={db.length}`  needs to be fixed to something else, but what could I use to get the .length of the filtered table? 

Oh yes, and why would the code breaks when I type in the input box?

**Comment by Elindo586** [comment]
I see an error now..  don't know how to fix it yet.. 

```
Unhandled Runtime Error
TypeError: val.Reference.toLowerCase is not a function
```

**Comment by icyJoseph** [comment]
Perhaps not the root of the error, but

```js
.filter((val) => {
                if (searchTerm == "") {
                  return val;
                } else if (
                  val.Reference.toLowerCase().includes(searchTerm.toLowerCase())
                ) {
                }
              })
```

Is not so good, use `===` and return `booleans/truthy/falsy` explicitly, for each branch you open, leaving blanks, might work, but it'll also be harder to read.

```js
.filter((val) => {
 // use `===` here
  if (searchTerm === "") {
    return val;
  } else if (val.Reference.toLowerCase().includes(searchTerm.toLowerCase())) {
    return true; // or something else?
  }
  return false; // or something else?
});

```

**Reply by icyJoseph** [reply]
You should also use `key` when rendering lists

```js
 .map((item) => {
                return (
                  <tr key={item.Reference}>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
              })}
```

**Comment by icyJoseph** [comment, marked as answer]
# Solution

The issue is that `Reference` from `val` is, sometimes, a `number`, and numbers don't have a `toLowerCase` method. Cast it to a string.

`${val.Reference}.toLowerCase()`, etc, the question is, why are we lowercasing, numbers? perhaps you want to lowercase the description?

The moment you type something the `if (searchTerm === "")` is skipped, and the faulty code kicks in, where a `toLowerCase` is invoked on  a `number`, which is not possible.

## Suggestion

Perhaps something like:

```js
            {currentTableData
              .filter((val) => {
                if (searchTerm === "") {
                  return val;
                }

                return (
                  val.Description.toLowerCase().includes(
                    searchTerm.toLowerCase()
                  ) || `${val.Reference}`.includes(searchTerm)
                );
              })
              .map((item) => {
                return (
                  <tr key={item.Reference}>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
              })}
```

### Edit

I got curious and looked at your data. `Reference` is both `strings` and `numbers`. The safest is to just cast them to `string` with `${val.Reference}`, and keep it simple.


### `.map`, `key` and `index`

It is generally, not a good idea, but if there's no other unique identifier for data, while mapping you can do this:

```js
     .map((item, index) => {
                return (
                  <tr key={index}>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
              }
```

I'd deter people from using the index, but for learning purposes, perhaps it is fine. `map` takes a callback which is called with three parameters, the current element, the index, and the entire list, see this link: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#parameters


## Edit 2 - Questions

> It kind of works...
> 
> There are a few things that might be related or outside of the search....
> 
> 1. It isn't shrinking the pagination size... might be due to `totalCount={db.length}`  which I need to think what to change it to so it will be according to the size of the search items...

It is a good idea to do:

```js
  const [currentPage, setCurrentPage] = useState(1);
  const [searchTerm, setSearchTerm] = useState("");

   const filteredDb = useMemo(
    () =>
      db.filter((val) => {
        if (searchTerm === "") {
          return val;
        }

        return (
          val.Description.toLowerCase().includes(searchTerm.toLowerCase()) ||
          `${val.Reference}`.includes(searchTerm)
        );
      }),
    [searchTerm]
  );

  const firstPageIndex = (currentPage - 1) * PageSize;
  const lastPageIndex = firstPageIndex + PageSize;

  const currentPageData = filteredDb.slice(firstPageIndex, lastPageIndex);
```

And then user `filteredTable` to `.map` over, and as input to `totalCount={filteredTable.length}`

```jsx
 filteredTable.map((item, index) => {
                return (
                  <tr key={index}>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
          })
```

> 2. I think it is searching only one column... I need to add some ||  to include all columns ...

Yeah, see above, you can do pretty complex filters!

> 3. I don't think I can (always) use Reference as a key, because some References might be repeated with some like "TBD", meaning "To Be Determined"...   the unique value should be Description, or I can try to reupload a new document including a unique id ....

Ah that's true, but then, it is highly recommended to try and slap a unique identifier to these things! So yes, re-upload the `JSON` file with unique ids! 

> **_Using unique keys is important for rendering performance, because when the key changes, the entire sub-tree of the `key-ed` element is re-rendered, whereas React could've re-used what it rendered before (because the item is the same)_**

**Reply by Elindo586** [reply]
For some reason including the  `totalCount={filteredTable.length}`  just erases the entire pagination bar...

Maybe because the .slice in the currentTableData code is still working with the entire db ?

**Reply by icyJoseph** [reply]
Ah yeah, it has to be filtered like this:

```js
  const [currentPage, setCurrentPage] = useState(1);
  const [searchTerm, setSearchTerm] = useState("");

  const filteredDb = db.filter((val) => {
    if (searchTerm === "") {
      return val;
    }

    return (
      val.Description.toLowerCase().includes(searchTerm.toLowerCase()) ||
      `${val.Reference}`.includes(searchTerm)
    );
  });

  const firstPageIndex = (currentPage - 1) * PageSize;
  const lastPageIndex = firstPageIndex + PageSize;

  const currentPageData = filteredDb.slice(firstPageIndex, lastPageIndex);

```

Good catch! And then you map over `currentPageData` and use `totalCount={filteredDb.length}`

**Reply by icyJoseph** [reply]
And you can memo the `filteredDb`, so that changing the page doesn't trigger a recalculation of the filteredDb, given that it only depends on the value of search term.

```js
 const filteredDb = useMemo(
    () =>
      db.filter((val) => {
        if (searchTerm === "") {
          return val;
        }

        return (
          val.Description.toLowerCase().includes(searchTerm.toLowerCase()) ||
          `${val.Reference}`.includes(searchTerm)
        );
      }),
    [searchTerm]
  );
```

**Reply by Elindo586** [reply]
Ok... its pretty much working now... other than a few questions if you don't mind.... (questions below the code)

I got the code like this:

** notice all of the` || ` in the search...   I also added `searchTerm` to the brackets in the `currentTableData `memo..

```
const filteredDb = useMemo(
    () =>
      db.filter((val) => {
        if (searchTerm === "") {
          return val;
        }
        return (
          val.Description.toLowerCase().includes(searchTerm.toLowerCase()) ||
          `${val.Description}`.includes(searchTerm) ||
          val.Replacements.toLowerCase().includes(searchTerm.toLowerCase()) ||
          `${val.Reference}`.includes(searchTerm) ||
          `${val.Replacements}`.includes(searchTerm)
        );
      }),
    [searchTerm]
  );

  const currentTableData = useMemo(() => {
    const firstPageIndex = (currentPage - 1) * PageSize;
    const lastPageIndex = firstPageIndex + PageSize;
    return filteredDb.slice(firstPageIndex, lastPageIndex);
  }, [currentPage, searchTerm]);

```
Questions....

If I do something like  
`val.Reference.toLowerCase().includes(searchTerm.toLowerCase()) ||`  <--- then I get `TypeError: val.Reference.toLowerCase is not a function`.... why do you think that happens?   Most of the references are numbers, but they also include some letters at times like ` "Reference": "1031032N",`     the idea is to catch letters and numbers on all columns....   could I do something like ``${val.Reference}`.includes(searchTerm) ||`  for each column and still catch letters? 


Then.... something that isn't a big deal.... 

Say if I go to the very last page and I do a search, the search works, but because I was in the last page it shows a blank page until I hit an arrow back to the first page...   do I need to create some sort of memo to update the currentPage from the SearchTerm?

**Reply by icyJoseph** [reply]
> val.Reference.toLowerCase().includes(searchTerm.toLowerCase()) || <--- then I get TypeError: val.Reference.toLowerCase is not a function.... why do you think that happens? Most of the references are numbers, but they also include some letters at times like "Reference": "1031032N", the idea is to catch letters and numbers on all columns.... could I do something like ``${val.Reference}.includes(searchTerm) || for each column and still catch letters?

Because when they are a number, there's just no toLowerCase method on numbers, I've been trying to explain this. Trying to do this throws an error, as you see. 

To be formal, there's ways to call methods on numbers, I don't want to confuse you with more syntax, so I'll skip it. However, even with that syntax, `toLowerCase` is not a method, and it will still throw an error. 

What we do here, as I said above, is cast the numbers to strings, and since strings are already strings, the casting won't affect them.

```js
let val = 2;
typeof val; // 'number'
typeof `${val}` // 'string'
let foo = 'foo';
typeof `${foo}` // 'string'
foo === `${foo}` // true
```

This is a bit of type handling in JavaScript, which isn't always pleasant. People often praise TypeScript because it would've caught this error at compile time and would've shown you how to fix it, from the get go.

So yeah, it is fine to do:

```js
`${val.Reference}`.includes(searchTerm)
```

Even better would be to force `Reference` to be a string, but perhaps the database that contains them is already mixed up.

> Say if I go to the very last page and I do a search, the search works, but because I was in the last page it shows a blank page until I hit an arrow back to the first page... do I need to create some sort of memo to update the currentPage from the SearchTerm?

I'd say, keep it simple, if the total pages changes, and current page is larger than that, set it to the max page.

Alternatively, keep it simple, and if searchTerm changes, reset pages to zero, but I don't know how you've structured your code.

**Reply by Elindo586** [reply]
Ok... I got it.... so some of the items are combinations of numbers and letters, or you can say letters and numbers.... If I do a search like ``${val.something}` `  it would return the value of something regardless if there are upper or lower cases letters in the items searched...

In the other column in which I have only letters I noticed that I do have to use the .toLowerCase to catch the search items...

**Reply by icyJoseph** [reply]
> Ok... I got it.... so some of the items are combinations of numbers and letters, or you can say letters and numbers.... 

Almost, it's not combinations on each item, but at sibling level if you will, take a look here, one reference is an actual integer number, you could do arithmetic with it:

<img width="446" alt="Screenshot 2022-07-14 at 19 53 31" src="https://user-images.githubusercontent.com/21013447/179050089-38230aa8-a467-432e-b10a-95f369893197.png">

This is one of the things with working with no-sql databases such as MongoDB, the Reference type has been made into a union of numbers and strings, and now the front end has to handle that.

**Reply by Elindo586** [reply]
I am proud to announce that I came up with my most complex code without using modification of an existing code...   this one is to fix the blank page when doing a search from the very last page...  it maintains the range of the current page the same as the range of the filtered items...

```
const updateFirstPage = useMemo(() => {
    const totalRange = filteredDb;
    if (currentPage > totalRange) return (currentPage = totalRange);
  }, [searchTerm]);

```

Thanks much to icyJoseph for helping me connect new brain cells!

**Reply by Elindo586** [reply]
Or not...  testing it..  it was working... now ... don't know..

**Reply by icyJoseph** [reply]
Maybe try with:

```js
 const totalPages = Math.ceil(filteredDb.length / PageSize);

  useEffect(() => {
    setCurrentPage((prev) =>
      prev > totalPages ? totalPages : Math.max(1, prev)
    );
  }, [totalPages]);
```

But I think this should be something the Pagination component controls. Anyway, don't forget to mark as answered, and ask away if you have more doubts!

**Reply by Elindo586** [reply]
So....

It works...

How can I get as good as you at coding?

**Comment by Elindo586** [comment]
I got the code below changed, and still doesn't work.....   maybe I can't call the .filter() method for currentTableData?    maybe it only work with an array? 


I am thinking.... if I find a way....  maybe I should filter the raw data, then pass a setNewData to the to the currentTableData.... ???     ... but I am not sure if it is possible to pass a setNewData state while using useMemo in currentTableData... 




```
<tbody>
            {currentTableData
              .filter((val) => {
                if (searchTerm === "") {
                  return val;
                } else if (
                  val.Reference.toLowerCase().includes(searchTerm.toLowerCase())
                ) {
                  return true;
                }
                return false;
              })
              .map((item) => {
                return (
                  <tr key={item.Reference}>
                    <td>{item.Reference}</td>
                    <td>{item.Description}</td>
                    <td>{item.Replacements}</td>
                  </tr>
                );
              })}
          </tbody>
        </table>
        <Pagination
          className="pagination-bar"
          currentPage={currentPage}
          totalCount={db.length}
          pageSize={PageSize}
          onPageChange={(page) => setCurrentPage(page)}
        />
      </div>


```

**Reply by icyJoseph** [reply]
I explained the solution in https://github.com/vercel/next.js/discussions/38618#discussioncomment-3140816, but to avoid further confusion I'll say here, that the way you have it `currentTableData` is an array, so `filter` is possible. See my other answer to fix the problem.

**Comment by Elindo586** [comment]
Oh.... BTW..... my data doesn't have an id to pass as a key... and some items might be repeated....  would that be a problem to set a key??????

With the old code doing a loop it was done like this...  but trying to change the code I don't have the "i" any more...
```
for (let i = itemStart; i <= itemEnd; i++) {
        if (partsList[i]) {
          itemsPerPage.push(
            <tr key={`tr${i}`}>
              <td key={`ref${i}`}>{partsList[i].Reference}</td>
              <td key={`desc${i}`}>{partsList[i].Description}</td>
              <td key={`repl${i}`}>{partsList[i].Replacements}</td>
            </tr>

```

**Reply by icyJoseph** [reply]
See my answer, titled `Solution`, https://github.com/vercel/next.js/discussions/38618#discussioncomment-3140816.

**Comment by Elindo586** [comment]
It kind of works...

There are a few things that might be related or outside of the search....

1.  It isn't shrinking the pagination size... might be due to `totalCount={db.length}`  which I need to think what to change it to so it will be according to the size of the search items...

2. I think it is searching only one column... I need to add some ||  to include all columns ...

3.  I don't think I can (always) use Reference as a key, because some References might be repeated with some like "TBD", meaning "To Be Determined"...   the unique value should be Description, or I can try to reupload a new document including a unique id ....

**Reply by icyJoseph** [reply]
Hi, I address this questions in https://github.com/vercel/next.js/discussions/38618#discussioncomment-3140816, please try to answer there as well.
